docs/JAVA_GRAMMAR_UPDATES.md

```diff
@@ -0,0 +1,225 @@
+  token types in our lexer grammar.
+- **Parser Rule**: A parser rule is a rule that defines the syntax of the
+  programming language. For example, a class declaration.
+- **Parser Rule Context**: A parser rule context is a context in which a parser
+  rule is applied, and contains information about the source code that matches
+  the parser rule. We use parser rule contexts to build our AST.
+- **Visitor**: A visitor is a design pattern that allows us to separate the
+  algorithm from the object structure. We use a visitor to traverse the parse
+  tree and build our AST.
+- **JavaAstVisitor**: A visitor that we use to traverse the parse tree and build
+  our AST. We define our
+  visitors [here](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/JavaAstVisitor.java).
+
+## Java Grammar Update Procedure
+
+Let's walk through an example of updating the Java grammar to support a new language feature. We
+will use the `when`
+expression as an example: https://openjdk.org/jeps/441.
+
+### Review the JEP and JLS
+
+It is good to first take some time to read the JEP to understand the new language feature.
+The JEP provides detailed information about the goals and motivations behind the new feature,
+And help us to come up with good testing strategies.
+
+It is also important to review the Java Language Specification (JLS) to understand the syntax
+and semantics of the new language feature. The JLS provides the formal definition of the Java
+programming language and is the ultimate reference for language features.
+
+The JLS defines the `when` expression as follows (surrounding context provided for clarity):
+
+```txt
+SwitchBlock:
+  { SwitchRule {SwitchRule} }
+  { {SwitchBlockStatementGroup} {SwitchLabel :} }
+
+SwitchRule:
+  SwitchLabel -> Expression ;
+  SwitchLabel -> Block
+  SwitchLabel -> ThrowStatement
+
+SwitchBlockStatementGroup:
+  SwitchLabel : {SwitchLabel :} BlockStatements
+
+SwitchLabel:
+  case CaseConstant {, CaseConstant}
+  case null [, default]
+  case CasePattern {, CasePattern} [Guard]
+  default
+
+CaseConstant:
+  ConditionalExpression
+
+CasePattern:
+  Pattern
+
+Guard:
+  when Expression
+```
+
+### Identify New Tokens and Update Lexer Grammar
+
+Not every new language feature requires new tokens. But in the case of the `when` expression,
+we need to introduce a new token to represent the `when` keyword. This requires an update
+to the lexer grammar.
+
+```antlr
+LITERAL_WHEN : 'when' ;
+```
+
+Notes:
+
+- We preface literal tokens with `LITERAL_` to distinguish them from other tokens.
+- Lexer rules are written in uppercase by convention.
+
+We will also need to add a
+new [TokenType](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java)
+in this case. However, we may not need to always add a token type for every new lexer token, it
+depends on the use case.
```

Comment:
Can we add some examples when we will and won't add them. For example, if we want to be able to define a token in our AST, it will need a type that will define when our checks stop on them in the tree. If there is no AST representation then we don't need a token type.

Reply:
Done, used SLIST as an example of a token used in two places, which demonstrates how we can "reuse" tokens.